kairos-contracts/demo-contract/contract/src/utils.rs

```diff
@@ -0,0 +1,38 @@
+/*
+    The utilities found in this file were scraped from other Casper contracts,
+    mainly cep-78 and cep-18.
+    This file is not necessarily due for review, unless breaking changes are suspected.
+*/
+use casper_contract::contract_api::runtime;
+use casper_types::{system::CallStackElement, Key};
+
+pub mod errors;
+pub mod events;
+use errors::DepositError;
+
+/// Wrap the immediate caller as a Key and return it
+fn call_stack_element_to_key(call_stack_element: CallStackElement) -> Key {
+    match call_stack_element {
+        CallStackElement::Session { account_hash } => Key::from(account_hash),
+        CallStackElement::StoredSession { account_hash, .. } => {
+            // Stored session code acts in account's context, so if stored session wants to interact
+            // with an CEP-18 token caller's address will be used.
+            Key::from(account_hash)
+        }
+        CallStackElement::StoredContract {
+            contract_package_hash,
+            ..
+        } => Key::from(contract_package_hash),
+    }
+}
+
+/// Traverse the callstack to retrieve the n - 1 th element of the callstack
+pub(crate) fn get_immediate_caller() -> Result<Key, DepositError> {
+    let call_stack = runtime::get_call_stack();
+    call_stack
+        .into_iter()
+        .rev()
+        .nth(1)
+        .map(call_stack_element_to_key)
+        .ok_or(DepositError::InvalidContext)
+}
```

Comment:
@jonas089 Can you explain this code to me.
I am unsure of the details.

Why `nth(1)`, what would the account in `nth(0)` be?
Can an attacker control this?
What is the expected total depth of this stack, does it not matter since `rev`?

Reply:
nth(0) when exeuting a contract would resolve to the contract that is being called itself rather than the instance (Account, Contract) interacting with it.

Reply:
Ok, I'm still concerned, but it's good enough for now.
We should revisit this in a security review at a much later date.